weave/gql_op_plugin.py

```diff
@@ -0,0 +1,65 @@
+import typing
+from dataclasses import dataclass
+
+from . import input_provider
+from . import op_args
+from . import op_def
+from . import gql_with_keys
+from . import input_provider
+from . import weave_types as types
+
+# A GQLKeyPropFn is a function that is called during the refinement phase of the compile pass
+# to propagate the GQL keys of a node's input types to its output type.
+GQLKeyPropFn = typing.Callable[[input_provider.InputProvider, types.Type], types.Type]
+
+
+@dataclass
+class GqlOpPlugin:
+    query_fn: typing.Callable[[input_provider.InputAndStitchProvider, str], str]
+    is_root: bool = False
+    root_resolver: typing.Optional["op_def.OpDef"] = None
+
+    # given the input types to the op, return a new output type with the input types'
+    # gql keys propagated appropriately. this is not a part of output_type to avoid
+    # the UI needing to make additional network requests to get the output type
+    gql_key_prop_fn: typing.Optional[GQLKeyPropFn] = None
+
+
+# Ops in `domain_ops` can add this plugin to their op definition to indicate
+# that they need data to be fetched from the GQL API. At it's core, the plugin
+# allows the user to specify a `query_fn` that takes in the inputs to the op and
+# returns a GQL query fragment that is needed by the calling op. The plugin also
+# allows the user to specify whether the op is a root op which indicates it is
+# the "top" of the GQL query tree. Note, while ops can use this directly (eg.
+# see `project_ops.py::artifacts`), most ops use the higher level helpers
+# defined in `wb_domain_gql.py`
+def wb_gql_op_plugin(
+    query_fn: typing.Callable[[input_provider.InputAndStitchProvider, str], str],
+    is_root: bool = False,
+    root_resolver: typing.Optional["op_def.OpDef"] = None,
+    gql_key_propagation_fn: typing.Optional[GQLKeyPropFn] = None,
```

Comment:
when is it ok for this to be None? Is it only when the gql op is a simple edge?

Reply:
Its always OK for this to be none. In some cases, it is still none when it could be implmemented, because I haven't finished implementing this for all of the #6 special business logic ops yet. 

When this is none, we just lose the key type propagation, and go back to effectively what we have now.